client/checkout/woopay/express-button/woopay-express-checkout-button.js

```diff
@@ -69,7 +79,141 @@ export const WoopayExpressCheckoutButton = ( {
 		}
 	}, [ isPreview, context ] );
 
-	const initWooPay = ( e ) => {
+	const newIframe = useCallback( () => {
+		if ( ! getConfig( 'isWoopayFirstPartyAuthEnabled' ) ) {
+			return;
+		}
+
+		const getWoopayOtpUrl = () => {
+			const tracksUserId = JSON.stringify(
+				getConfig( 'tracksUserIdentity' )
+			);
+
+			const urlParams = new URLSearchParams();
+			urlParams.append( 'testMode', getConfig( 'testMode' ) );
+			urlParams.append( 'source_url', window.location.href );
+			urlParams.append( 'tracksUserIdentity', tracksUserId );
+
+			return getConfig( 'woopayHost' ) + '/otp/?' + urlParams.toString();
+		};
+
+		const iframe = document.createElement( 'iframe' );
+		iframe.src = getWoopayOtpUrl();
+		iframe.height = 0;
+		iframe.style.visibility = 'hidden';
+		iframe.style.position = 'absolute';
+		iframe.style.top = '0';
+
+		iframe.addEventListener( 'load', () => {
+			// Change button's onClick handle to use express checkout flow.
+			initWoopayRef.current = ( e ) => {
+				e.preventDefault();
+
+				if ( isPreview ) {
+					return;
+				}
+
+				wcpayTracks.recordUserEvent(
+					wcpayTracks.events.WOOPAY_BUTTON_CLICK,
+					{
+						source: context,
+					}
+				);
+
+				if ( isProductPage ) {
+					const productData = getProductDataRef.current();
+
+					if ( ! productData ) {
+						return;
+					}
+
+					if ( listenForCartChanges.stop ) {
+						// Temporarily stop listening for cart changes to prevent
+						// rendering a new button + iFrame when the cart is updated.
+						listenForCartChanges.stop();
+					}
+
+					addToCartRef.current( productData ).then( () => {
+						if ( listenForCartChanges.start ) {
+							// Start listening for cart changes, again.
+							listenForCartChanges.start();
+						}
+						request(
+							buildAjaxURL(
+								getConfig( 'wcAjaxUrl' ),
+								'get_woopay_session'
+							),
+							{
+								_ajax_nonce: getConfig( 'woopaySessionNonce' ),
+							}
+						).then( ( response ) => {
```

Comment:
Is there any side effects in case this request fails? Either on a 2xx or a 4xx status.

Reply:
Good callout. Before, on a 400 HTTP response, the button would fail silently. This has been fixed in https://github.com/Automattic/woocommerce-payments/pull/7199/commits/a9e994aed5191dc02d193778b07e06be712b3174 and https://github.com/Automattic/woocommerce-payments/pull/7199/commits/6e55eda8ebe9045d7b5b307bfcaa19b5f541c887. Now, an error notification appears.